--- a/targets/targets.mk
+++ b/targets/targets.mk
@@ -16,6 +16,7 @@ $(eval $(call GluonTarget,ramips,mt7620))
 $(eval $(call GluonTarget,ramips,mt7621))
 $(eval $(call GluonTarget,ramips,mt76x8))
 $(eval $(call GluonTarget,ramips,rt305x))
+$(eval $(call GluonTarget,rockchip,armv8))
 $(eval $(call GluonTarget,sunxi,cortexa7))
 $(eval $(call GluonTarget,x86,generic))
 $(eval $(call GluonTarget,x86,geode))
