diff --git a/modules b/modules
index 02d4826f..0b535a55 100644
--- a/modules
+++ b/modules
@@ -1,12 +1,12 @@
 GLUON_FEEDS='packages routing gluon'
 
-OPENWRT_REPO=https://git.openwrt.org/openwrt/openwrt.git
+OPENWRT_REPO=https://github.com/openwrt/openwrt.git
 OPENWRT_BRANCH=openwrt-19.07
-OPENWRT_COMMIT=5af8da37870dc05dbe2e57e04be714b80f4aa21d
+OPENWRT_COMMIT=29b4104d69bf91db17764dd885e9e111a373f08c
 
 PACKAGES_PACKAGES_REPO=https://github.com/openwrt/packages.git
 PACKAGES_PACKAGES_BRANCH=openwrt-19.07
-PACKAGES_PACKAGES_COMMIT=59d39c09d84fb08675cc58d4ec32837e9163b017
+PACKAGES_PACKAGES_COMMIT=a2673dc53c4689798c1d70d7342cb3efadb0af74
 
 PACKAGES_ROUTING_REPO=https://github.com/openwrt-routing/packages.git
 PACKAGES_ROUTING_BRANCH=openwrt-19.07
diff --git a/patches/openwrt/0009-ath79-enable-GL-AR750S-NOR-variant-from-master.patch b/patches/openwrt/0009-ath79-enable-GL-AR750S-NOR-variant-from-master.patch
new file mode 100644
index 00000000..1b9848d7
--- /dev/null
+++ b/patches/openwrt/0009-ath79-enable-GL-AR750S-NOR-variant-from-master.patch
@@ -0,0 +1,61 @@
+From: Jan Alexander <jan@nalx.net>
+Date: Tue, 31 Mar 2020 21:50:28 +0200
+Subject: ath79: enable GL-AR750S NOR variant from master
+
+diff --git a/target/linux/ath79/base-files/etc/board.d/02_network b/target/linux/ath79/base-files/etc/board.d/02_network
+index 5dda551caae0429880ee9d5965bfb6797d218e6d..b8fac8816c9a2b2a87a5d1335b41127666afe2e4 100755
+--- a/target/linux/ath79/base-files/etc/board.d/02_network
++++ b/target/linux/ath79/base-files/etc/board.d/02_network
+@@ -155,7 +155,7 @@ ath79_setup_interfaces()
+ 	etactica,eg200)
+ 		ucidef_set_interface_lan "eth0" "dhcp"
+ 		;;
+-	glinet,gl-ar750s)
++	glinet,gl-ar750s-nor)
+ 		ucidef_add_switch "switch0" \
+ 			"0@eth0" "2:lan:2" "3:lan:1" "1:wan"
+ 		;;
+diff --git a/target/linux/ath79/base-files/etc/hotplug.d/firmware/11-ath10k-caldata b/target/linux/ath79/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
+index d93e6dcd71ab19c53905daa41e95cc4fc614f114..c917f38211d0b246f064dba4b7feefecf61f5856 100644
+--- a/target/linux/ath79/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
++++ b/target/linux/ath79/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
+@@ -117,7 +117,7 @@ case "$FIRMWARE" in
+ 		ath10kcal_extract "art" 20480 2116
+ 		ath10kcal_patch_mac $(macaddr_add $(cat /sys/class/net/eth0/address) +1)
+ 		;;
+-	glinet,gl-ar750s)
++	glinet,gl-ar750s-nor)
+ 		ath10kcal_extract "art" 20480 2116
+ 		ath10kcal_patch_mac $(macaddr_add $(mtd_get_mac_binary art 0) +1)
+ 		;;
+diff --git a/target/linux/ath79/dts/qca9563_glinet_gl-ar750s.dts b/target/linux/ath79/dts/qca9563_glinet_gl-ar750s.dts
+index 03922bcd1fe9a453d5916537609317b94eea18c6..ff64e16d1ce7a94d16529e5954e1d50513a5e2cb 100644
+--- a/target/linux/ath79/dts/qca9563_glinet_gl-ar750s.dts
++++ b/target/linux/ath79/dts/qca9563_glinet_gl-ar750s.dts
+@@ -7,8 +7,8 @@
+ #include "qca956x.dtsi"
+ 
+ / {
+-	compatible = "glinet,gl-ar750s", "qca,qca9563";
+-	model = "GL.iNet GL-AR750S";
++	compatible = "glinet,gl-ar750s-nor", "qca,qca9563";
++	model = "GL.iNet GL-AR750S (NOR)";
+ 
+ 	aliases {
+ 		led-boot = &power;
+diff --git a/target/linux/ath79/image/generic.mk b/target/linux/ath79/image/generic.mk
+index 55053be34f11f0df982c85f94c9180fdba9ff221..892ef10f870e347c8a1509cecd35bce4b5e98bee 100644
+--- a/target/linux/ath79/image/generic.mk
++++ b/target/linux/ath79/image/generic.mk
+@@ -403,9 +403,9 @@ define Device/glinet_gl-ar750s
+   DEVICE_TITLE := GL.iNet GL-AR750S
+   DEVICE_PACKAGES := kmod-usb2 kmod-ath10k-ct ath10k-firmware-qca9887-ct block-mount
+   IMAGE_SIZE := 16000k
+-  SUPPORTED_DEVICES += gl-ar750s
++  SUPPORTED_DEVICES += gl-ar750s glinet,gl-ar750s-nor
+ endef
+-#TARGET_DEVICES += glinet_gl-ar750s
++TARGET_DEVICES += glinet_gl-ar750s
+ 
+ define Device/glinet_gl-x750
+   ATH_SOC := qca9531
diff --git a/patches/openwrt/0010-tools-add-zstd.patch b/patches/openwrt/0010-tools-add-zstd.patch
new file mode 100644
index 00000000..7862dce7
--- /dev/null
+++ b/patches/openwrt/0010-tools-add-zstd.patch
@@ -0,0 +1,113 @@
+From: Matthias Schiffer <mschiffer@universe-factory.net>
+Date: Wed, 13 May 2020 20:22:12 +0200
+Subject: tools: add zstd
+
+Signed-off-by: Matthias Schiffer <mschiffer@universe-factory.net>
+(cherry picked from commit 258dc0d0fd3aae47add9b7dca40848a92d03a4ea)
+
+diff --git a/tools/Makefile b/tools/Makefile
+index d7207ba89dd91df558eaf970961fdef225aa1f37..14fe4fb4b5f4e0c745cb8592a39bcf238dcc5444 100644
+--- a/tools/Makefile
++++ b/tools/Makefile
+@@ -33,7 +33,7 @@ tools-$(CONFIG_TARGET_mxs) += elftosb sdimage
+ tools-$(CONFIG_TARGET_ar71xx) += lzma-old
+ tools-$(CONFIG_TARGET_ar71xx)$(CONFIG_TARGET_ath79) += squashfs
+ tools-$(CONFIG_USES_MINOR) += kernel2minor
+-tools-y += lzma squashfskit4 zip
++tools-y += lzma squashfskit4 zip zstd
+ tools-$(BUILD_B43_TOOLS) += b43-tools
+ tools-$(BUILD_ISL) += isl
+ tools-$(CONFIG_USE_SPARSE) += sparse
+diff --git a/tools/zstd/Makefile b/tools/zstd/Makefile
+new file mode 100644
+index 0000000000000000000000000000000000000000..7459725e8e79b846ed96551753d07fdd02459598
+--- /dev/null
++++ b/tools/zstd/Makefile
+@@ -0,0 +1,20 @@
++include $(TOPDIR)/rules.mk
++
++PKG_NAME:=zstd
++PKG_VERSION:=1.4.4
++
++PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
++PKG_SOURCE_URL:=@GITHUB/facebook/zstd/releases/download/v$(PKG_VERSION)
++PKG_HASH:=a364f5162c7d1a455cc915e8e3cf5f4bd8b75d09bc0f53965b0c9ca1383c52c8
++
++PKG_LICENSE:=BSD-3-Clause
++PKG_LICENSE_FILES:=LICENSE
++PKG_CPE_ID:=cpe:/a:facebook:zstandard
++
++HOST_BUILD_PARALLEL:=1
++
++include $(INCLUDE_DIR)/host-build.mk
++
++HOST_MAKE_FLAGS = PREFIX=$(HOST_BUILD_PREFIX) HAVE_ZLIB=0 HAVE_LZMA=0 HAVE_LZ4=0
++
++$(eval $(call HostBuild))
+diff --git a/tools/zstd/patches/0001-build-issue-More-portable-header-prefix-usage-1987.patch b/tools/zstd/patches/0001-build-issue-More-portable-header-prefix-usage-1987.patch
+new file mode 100644
+index 0000000000000000000000000000000000000000..6d743aa3855262c2a0956f70ec99588ce9ebe53b
+--- /dev/null
++++ b/tools/zstd/patches/0001-build-issue-More-portable-header-prefix-usage-1987.patch
+@@ -0,0 +1,61 @@
++From 06a57cf57e3c4e887cadcf688e3081154f3f6db4 Mon Sep 17 00:00:00 2001
++Message-Id: <06a57cf57e3c4e887cadcf688e3081154f3f6db4.1589392463.git.mschiffer@universe-factory.net>
++From: Bimba Shrestha <bimbashrestha@fb.com>
++Date: Thu, 6 Feb 2020 14:10:51 -0800
++Subject: [PATCH] [build-issue] More portable header prefix usage (#) (#1987)
++
++* make 4.3 build issue fix
++
++* Changing header name and adding comment
++---
++ programs/Makefile | 11 +++++++----
++ 1 file changed, 7 insertions(+), 4 deletions(-)
++
++diff --git a/programs/Makefile b/programs/Makefile
++index b75314a83f43..a9ee3cb5311b 100644
++--- a/programs/Makefile
+++++ b/programs/Makefile
++@@ -94,9 +94,12 @@ endif
++ 
++ VOID = /dev/null
++ 
+++# Make 4.3 doesn't support '\#' anymore (https://lwn.net/Articles/810071/)
+++NUM_SYMBOL := \#
+++
++ # thread detection
++ NO_THREAD_MSG := ==> no threads, building without multithreading support
++-HAVE_PTHREAD := $(shell printf '\#include <pthread.h>\nint main(void) { return 0; }' > have_pthread.c && $(CC) $(FLAGS) -o have_pthread$(EXT) have_pthread.c -pthread 2> $(VOID) && rm have_pthread$(EXT) && echo 1 || echo 0; rm have_pthread.c)
+++HAVE_PTHREAD := $(shell printf '$(NUM_SYMBOL)include <pthread.h>\nint main(void) { return 0; }' > have_pthread.c && $(CC) $(FLAGS) -o have_pthread$(EXT) have_pthread.c -pthread 2> $(VOID) && rm have_pthread$(EXT) && echo 1 || echo 0; rm have_pthread.c)
++ HAVE_THREAD := $(shell [ "$(HAVE_PTHREAD)" -eq "1" -o -n "$(filter Windows%,$(OS))" ] && echo 1 || echo 0)
++ ifeq ($(HAVE_THREAD), 1)
++ THREAD_MSG := ==> building with threading support
++@@ -108,7 +111,7 @@ endif
++ 
++ # zlib detection
++ NO_ZLIB_MSG := ==> no zlib, building zstd without .gz support
++-HAVE_ZLIB := $(shell printf '\#include <zlib.h>\nint main(void) { return 0; }' > have_zlib.c && $(CC) $(FLAGS) -o have_zlib$(EXT) have_zlib.c -lz 2> $(VOID) && rm have_zlib$(EXT) && echo 1 || echo 0; rm have_zlib.c)
+++HAVE_ZLIB := $(shell printf '$(NUM_SYMBOL)include <zlib.h>\nint main(void) { return 0; }' > have_zlib.c && $(CC) $(FLAGS) -o have_zlib$(EXT) have_zlib.c -lz 2> $(VOID) && rm have_zlib$(EXT) && echo 1 || echo 0; rm have_zlib.c)
++ ifeq ($(HAVE_ZLIB), 1)
++ ZLIB_MSG := ==> building zstd with .gz compression support
++ ZLIBCPP = -DZSTD_GZCOMPRESS -DZSTD_GZDECOMPRESS
++@@ -119,7 +122,7 @@ endif
++ 
++ # lzma detection
++ NO_LZMA_MSG := ==> no liblzma, building zstd without .xz/.lzma support
++-HAVE_LZMA := $(shell printf '\#include <lzma.h>\nint main(void) { return 0; }' > have_lzma.c && $(CC) $(FLAGS) -o have_lzma$(EXT) have_lzma.c -llzma 2> $(VOID) && rm have_lzma$(EXT) && echo 1 || echo 0; rm have_lzma.c)
+++HAVE_LZMA := $(shell printf '$(NUM_SYMBOL)include <lzma.h>\nint main(void) { return 0; }' > have_lzma.c && $(CC) $(FLAGS) -o have_lzma$(EXT) have_lzma.c -llzma 2> $(VOID) && rm have_lzma$(EXT) && echo 1 || echo 0; rm have_lzma.c)
++ ifeq ($(HAVE_LZMA), 1)
++ LZMA_MSG := ==> building zstd with .xz/.lzma compression support
++ LZMACPP = -DZSTD_LZMACOMPRESS -DZSTD_LZMADECOMPRESS
++@@ -130,7 +133,7 @@ endif
++ 
++ # lz4 detection
++ NO_LZ4_MSG := ==> no liblz4, building zstd without .lz4 support
++-HAVE_LZ4 := $(shell printf '\#include <lz4frame.h>\n\#include <lz4.h>\nint main(void) { return 0; }' > have_lz4.c && $(CC) $(FLAGS) -o have_lz4$(EXT) have_lz4.c -llz4 2> $(VOID) && rm have_lz4$(EXT) && echo 1 || echo 0; rm have_lz4.c)
+++HAVE_LZ4 := $(shell printf '$(NUM_SYMBOL)include <lz4frame.h>\n\#include <lz4.h>\nint main(void) { return 0; }' > have_lz4.c && $(CC) $(FLAGS) -o have_lz4$(EXT) have_lz4.c -llz4 2> $(VOID) && rm have_lz4$(EXT) && echo 1 || echo 0; rm have_lz4.c)
++ ifeq ($(HAVE_LZ4), 1)
++ LZ4_MSG := ==> building zstd with .lz4 compression support
++ LZ4CPP = -DZSTD_LZ4COMPRESS -DZSTD_LZ4DECOMPRESS
++-- 
++2.26.2
++
diff --git a/patches/openwrt/0011-build-compress-kernel-debuginfo-using-zstd.patch b/patches/openwrt/0011-build-compress-kernel-debuginfo-using-zstd.patch
new file mode 100644
index 00000000..66678df2
--- /dev/null
+++ b/patches/openwrt/0011-build-compress-kernel-debuginfo-using-zstd.patch
@@ -0,0 +1,43 @@
+From: Matthias Schiffer <mschiffer@universe-factory.net>
+Date: Wed, 13 May 2020 20:33:46 +0200
+Subject: build: compress kernel debuginfo using zstd
+
+zstd with its default settings (compression level -3) compresses better
+than bzip2 -9 (which is the default setting), and is an order of magnitude
+faster.
+
+I made the following measurements for the most common compression tools
+(all standard Debian Buster versions, default flags unless noted
+otherwise), using the debug information of a large x86-64 kernel with
+ALL_KMODS:
+
+* kernel-debug.tar: 376M
+* kernel-debug.tar.gz: 101M, compressed in ~12s
+* kernel-debug.tar.bz2: 91M, compressed in ~15s
+* kernel-debug.tar.xz: 57M, compressed in ~101s
+* kernel-debug.tar.zst: 86M, compressed in ~1s
+
+With zstd, there is still some room for improvement by increasing the
+compression, but the slight increase in compression ratio
+(22.83% -> 19.46%) does not justify the significant increase in
+compression time (about 5 times on my machine) in my opinion.
+
+Note that multithreaded compression (-T argument) does not affect
+reproducibility with zstd.
+
+Signed-off-by: Matthias Schiffer <mschiffer@universe-factory.net>
+(cherry picked from commit 4bd7990488b0ca7b5cae16f0a9147a4146759053)
+
+diff --git a/include/kernel-build.mk b/include/kernel-build.mk
+index 3fdf7efc52857a5184348cc1261848f75751b8a9..af7c3a8f0bb15c7e0d7072876705ff0bf4f9c8d1 100644
+--- a/include/kernel-build.mk
++++ b/include/kernel-build.mk
+@@ -70,7 +70,7 @@ ifdef CONFIG_COLLECT_KERNEL_DEBUG
+ 	$(FIND) $(KERNEL_BUILD_DIR)/debug -type f | $(XARGS) $(KERNEL_CROSS)strip --only-keep-debug
+ 	$(TAR) c -C $(KERNEL_BUILD_DIR) debug \
+ 		$(if $(SOURCE_DATE_EPOCH),--mtime="@$(SOURCE_DATE_EPOCH)") \
+-		| bzip2 -c -9 > $(BIN_DIR)/kernel-debug.tar.bz2
++		| zstd -T0 -f -o $(BIN_DIR)/kernel-debug.tar.zst
+   endef
+ endif
+ 
diff --git a/patches/openwrt/0012-mac80211-rt2800-enable-MFP-support-unconditionally.patch b/patches/openwrt/0012-mac80211-rt2800-enable-MFP-support-unconditionally.patch
new file mode 100644
index 00000000..0a8ca213
--- /dev/null
+++ b/patches/openwrt/0012-mac80211-rt2800-enable-MFP-support-unconditionally.patch
@@ -0,0 +1,62 @@
+From: Rui Salvaterra <rsalvaterra@gmail.com>
+Date: Mon, 25 May 2020 14:49:07 +0100
+Subject: mac80211: rt2800: enable MFP support unconditionally
+
+This gives us WPA3 support out of the box without having to manually disable
+hardware crypto. The driver will fall back to software crypto if the connection
+requires management frame protection.
+
+Signed-off-by: Daniel Golle <daniel@makrotopia.org>
+[apply to openwrt-1907]
+Signed-off-by: David Bauer <mail@david-bauer.net>
+
+diff --git a/package/kernel/mac80211/patches/rt2x00/080-rt2800-enable-MFP-support-unconditionally.patch b/package/kernel/mac80211/patches/rt2x00/080-rt2800-enable-MFP-support-unconditionally.patch
+new file mode 100644
+index 0000000000000000000000000000000000000000..1d55b2756c365a13b2315e874809e2397fb35855
+--- /dev/null
++++ b/package/kernel/mac80211/patches/rt2x00/080-rt2800-enable-MFP-support-unconditionally.patch
+@@ -0,0 +1,44 @@
++From b6b15e20421fefae9f78274f9fef80bc97bf5d5c Mon Sep 17 00:00:00 2001
++From: Rui Salvaterra <rsalvaterra@gmail.com>
++Date: Mon, 25 May 2020 14:49:07 +0100
++Subject: [PATCH] rt2800: enable MFP support unconditionally
++
++This gives us WPA3 support out of the box without having to manually disable
++hardware crypto. The driver will fall back to software crypto if the connection
++requires management frame protection.
++
++Suggested-by: Stanislaw Gruszka <stf_xl@wp.pl>
++Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>
++Acked-by: Stanislaw Gruszka <stf_xl@wp.pl>
++Signed-off-by: Kalle Valo <kvalo@codeaurora.org>
++Link: https://lore.kernel.org/r/20200525134906.1672-1-rsalvaterra@gmail.com
++---
++ drivers/net/wireless/ralink/rt2x00/rt2800lib.c | 4 +---
++ drivers/net/wireless/ralink/rt2x00/rt2x00mac.c | 3 ++-
++ 2 files changed, 3 insertions(+), 4 deletions(-)
++
++--- a/drivers/net/wireless/ralink/rt2x00/rt2800lib.c
+++++ b/drivers/net/wireless/ralink/rt2x00/rt2800lib.c
++@@ -9985,9 +9985,7 @@ static int rt2800_probe_hw_mode(struct r
++ 	if (!rt2x00_is_usb(rt2x00dev))
++ 		ieee80211_hw_set(rt2x00dev->hw, HOST_BROADCAST_PS_BUFFERING);
++ 
++-	/* Set MFP if HW crypto is disabled. */
++-	if (rt2800_hwcrypt_disabled(rt2x00dev))
++-		ieee80211_hw_set(rt2x00dev->hw, MFP_CAPABLE);
+++	ieee80211_hw_set(rt2x00dev->hw, MFP_CAPABLE);
++ 
++ 	SET_IEEE80211_DEV(rt2x00dev->hw, rt2x00dev->dev);
++ 	SET_IEEE80211_PERM_ADDR(rt2x00dev->hw,
++--- a/drivers/net/wireless/ralink/rt2x00/rt2x00mac.c
+++++ b/drivers/net/wireless/ralink/rt2x00/rt2x00mac.c
++@@ -459,7 +459,8 @@ int rt2x00mac_set_key(struct ieee80211_h
++ 	if (!test_bit(DEVICE_STATE_PRESENT, &rt2x00dev->flags))
++ 		return 0;
++ 
++-	if (!rt2x00_has_cap_hw_crypto(rt2x00dev))
+++	/* The hardware can't do MFP */
+++	if (!rt2x00_has_cap_hw_crypto(rt2x00dev) || (sta && sta->mfp))
++ 		return -EOPNOTSUPP;
++ 
++ 	/*
diff --git a/patches/openwrt/0013-mt76-mt76x0-disable-GTK-offloading.patch b/patches/openwrt/0013-mt76-mt76x0-disable-GTK-offloading.patch
new file mode 100644
index 00000000..c1f404c8
--- /dev/null
+++ b/patches/openwrt/0013-mt76-mt76x0-disable-GTK-offloading.patch
@@ -0,0 +1,49 @@
+From: David Bauer <mail@david-bauer.net>
+Date: Sat, 13 Jun 2020 19:19:17 +0200
+Subject: mt76: mt76x0: disable GTK offloading
+
+When the GTK is offloaded, MT7610 won't transmit any multicast frames.
+This is most likely due to a bug in the offloading datapath. MT7612 is
+not affected.
+
+Disable GTK offloading for now. It can be re-enabled once the bug in the
+offloading path is fixed.
+
+Signed-off-by: David Bauer <mail@david-bauer.net>
+
+diff --git a/package/kernel/mt76/patches/001-mt76-mt76x0-disable-gtk-offloading.patch b/package/kernel/mt76/patches/001-mt76-mt76x0-disable-gtk-offloading.patch
+new file mode 100644
+index 0000000000000000000000000000000000000000..e7e19ac957dbfaa9510016d3387abe9eed353538
+--- /dev/null
++++ b/package/kernel/mt76/patches/001-mt76-mt76x0-disable-gtk-offloading.patch
+@@ -0,0 +1,30 @@
++From ae01717951013fbc8bb0315d902d5b9f5873631a Mon Sep 17 00:00:00 2001
++From: David Bauer <mail@david-bauer.net>
++Date: Fri, 12 Jun 2020 01:09:57 +0200
++Subject: [PATCH] mt76: mt76x0: disable GTK offloading
++
++When the GTK is offloaded, MT7610 won't transmit any multicast frames.
++This is most likely due to a bug in the offloading datapath. MT7612 is
++not affected.
++
++Disable GTK offloading for now. It can be re-enabled once the bug in the
++offloading path is fixed.
++
++Signed-off-by: David Bauer <mail@david-bauer.net>
++---
++ drivers/net/wireless/mediatek/mt76/mt76x02_util.c | 4 ++++
++ 1 file changed, 4 insertions(+)
++
++--- a/mt76x02_util.c
+++++ b/mt76x02_util.c
++@@ -432,6 +432,10 @@ int mt76x02_set_key(struct ieee80211_hw
++ 	    !(key->flags & IEEE80211_KEY_FLAG_PAIRWISE))
++ 		return -EOPNOTSUPP;
++ 
+++	/* MT76x0 GTK offloading is currently broken */
+++	if (is_mt76x0(dev) && !(key->flags & IEEE80211_KEY_FLAG_PAIRWISE))
+++		return -EOPNOTSUPP;
+++
++ 	/*
++ 	 * In USB AP mode, broadcast/multicast frames are setup in beacon
++ 	 * data registers and sent via HW beacons engine, they require to
diff --git a/patches/packages/packages/0001-fastd-update-to-v19.patch b/patches/packages/packages/0001-fastd-update-to-v19.patch
new file mode 100644
index 00000000..8f35f478
--- /dev/null
+++ b/patches/packages/packages/0001-fastd-update-to-v19.patch
@@ -0,0 +1,164 @@
+From: Matthias Schiffer <mschiffer@universe-factory.net>
+Date: Fri, 22 May 2020 21:09:21 +0200
+Subject: fastd: update to v19
+
+Signed-off-by: Matthias Schiffer <mschiffer@universe-factory.net>
+
+diff --git a/net/fastd/Config.in b/net/fastd/Config.in
+index 3350eb3099a26c870d70373c0712a8b59881ee5c..e6440075e561093c86543943cb982d010a4ef0e0 100644
+--- a/net/fastd/Config.in
++++ b/net/fastd/Config.in
+@@ -36,16 +36,6 @@ config FASTD_ENABLE_METHOD_NULL
+ 	depends on PACKAGE_fastd
+ 	default y
+ 
+-config FASTD_ENABLE_METHOD_XSALSA20_POLY1305
+-	bool "Enable xsalsa20-poly1305 method"
+-	depends on PACKAGE_fastd
+-	default n
+-
+-
+-config FASTD_ENABLE_CIPHER_AES128_CTR
+-	bool "Enable the AES128-CTR cipher"
+-	depends on PACKAGE_fastd
+-	default n
+ 
+ config FASTD_ENABLE_CIPHER_NULL
+ 	bool "Enable the null cipher"
+diff --git a/net/fastd/Makefile b/net/fastd/Makefile
+index f4890b56931a75849229d25fe78720e19d493383..7483e7b003041fb59991d72d0ccfcc8a28bb17a3 100644
+--- a/net/fastd/Makefile
++++ b/net/fastd/Makefile
+@@ -8,13 +8,13 @@
+ include $(TOPDIR)/rules.mk
+ 
+ PKG_NAME:=fastd
+-PKG_VERSION:=18
+-PKG_RELEASE:=5
++PKG_VERSION:=19
++PKG_RELEASE:=2
+ 
+ PKG_MAINTAINER:=Matthias Schiffer <mschiffer@universe-factory.net>
+ PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+ PKG_SOURCE_URL:=https://github.com/NeoRaider/fastd/releases/download/v$(PKG_VERSION)
+-PKG_HASH:=714ff09d7bd75f79783f744f6f8c5af2fe456c8cf876feaa704c205a73e043c9
++PKG_HASH:=6054608e2103b634c9d19ecd1ae058d4ec694747047130719db180578729783a
+ 
+ PKG_LICENSE:=BSD-2-Clause
+ PKG_LICENSE_FILES:=COPYRIGHT
+@@ -27,8 +27,6 @@ PKG_CONFIG_DEPENDS:=\
+ 	CONFIG_FASTD_ENABLE_METHOD_GENERIC_POLY1305 \
+ 	CONFIG_FASTD_ENABLE_METHOD_GENERIC_UMAC \
+ 	CONFIG_FASTD_ENABLE_METHOD_NULL \
+-	CONFIG_FASTD_ENABLE_METHOD_XSALSA20_POLY1305 \
+-	CONFIG_FASTD_ENABLE_CIPHER_AES128_CTR \
+ 	CONFIG_FASTD_ENABLE_CIPHER_NULL \
+ 	CONFIG_FASTD_ENABLE_CIPHER_SALSA20 \
+ 	CONFIG_FASTD_ENABLE_CIPHER_SALSA2012 \
+@@ -44,6 +42,7 @@ PKG_CONFIG_DEPENDS:=\
+ 
+ 
+ PKG_BUILD_DEPENDS:=nacl
++PKG_BUILD_PARALLEL:=1
+ 
+ include $(INCLUDE_DIR)/package.mk
+ include $(INCLUDE_DIR)/cmake.mk
+@@ -73,7 +72,6 @@ CMAKE_OPTIONS += \
+ 	-DWITH_METHOD_GENERIC_POLY1305:BOOL=FALSE \
+ 	-DWITH_METHOD_GENERIC_UMAC:BOOL=FALSE \
+ 	-DWITH_METHOD_NULL:BOOL=FALSE \
+-	-DWITH_METHOD_XSALSA20_POLY1305:BOOL=FALSE \
+ 	-DWITH_CIPHER_AES128_CTR:BOOL=FALSE \
+ 	-DWITH_CIPHER_NULL:BOOL=FALSE \
+ 	-DWITH_CIPHER_SALSA20:BOOL=FALSE \
+@@ -120,14 +118,6 @@ ifeq ($(CONFIG_FASTD_ENABLE_METHOD_NULL),y)
+ CMAKE_OPTIONS += -DWITH_METHOD_NULL:BOOL=TRUE
+ endif
+ 
+-ifeq ($(CONFIG_FASTD_ENABLE_METHOD_XSALSA20_POLY1305),y)
+-CMAKE_OPTIONS += -DWITH_METHOD_XSALSA20_POLY1305:BOOL=TRUE
+-endif
+-
+-
+-ifeq ($(CONFIG_FASTD_ENABLE_CIPHER_AES128_CTR),y)
+-CMAKE_OPTIONS += -DWITH_CIPHER_AES128_CTR:BOOL=TRUE
+-endif
+ 
+ ifeq ($(CONFIG_FASTD_ENABLE_CIPHER_NULL),y)
+ CMAKE_OPTIONS += -DWITH_CIPHER_NULL:BOOL=TRUE
+diff --git a/net/fastd/patches/0001-resolve-fix-segmentation-fault-with-musl-1.1.20.patch b/net/fastd/patches/0001-resolve-fix-segmentation-fault-with-musl-1.1.20.patch
+deleted file mode 100644
+index 52c19174083c29e5da02cabb2ddc02474cf11b37..0000000000000000000000000000000000000000
+--- a/net/fastd/patches/0001-resolve-fix-segmentation-fault-with-musl-1.1.20.patch
++++ /dev/null
+@@ -1,35 +0,0 @@
+-From 9710132c04cd378bd36f16a2a3d98d9c4c5fdbac Mon Sep 17 00:00:00 2001
+-From: David Bauer <mail@david-bauer.net>
+-Date: Thu, 25 Jul 2019 18:51:25 +0200
+-Subject: [PATCH] resolve: fix segmentation fault with musl >1.1.20
+-
+-When compiled with musl >1.1.20, fastd will crash in case it can't
+-resolve a peers hostname. This is due to a changed implementation of
+-freeaddrinfo in musl 1.1.21 onwards.
+-
+-This segfault is fixed by not calling freeaddrinfo in case the supplied
+-pointer is null.
+-
+-Signed-off-by: David Bauer <mail@david-bauer.net>
+----
+- src/resolve.c | 4 +++-
+- 1 file changed, 3 insertions(+), 1 deletion(-)
+-
+-diff --git a/src/resolve.c b/src/resolve.c
+-index 9bdfa1c..bfd2a59 100644
+---- a/src/resolve.c
+-+++ b/src/resolve.c
+-@@ -104,7 +104,9 @@ static void * resolve_peer(void *varg) {
+- 
+- 	fastd_async_enqueue(ASYNC_TYPE_RESOLVE_RETURN, ret, sizeof(fastd_async_resolve_return_t) + n_addr*sizeof(fastd_peer_address_t));
+- 
+--	freeaddrinfo(res);
+-+	if (res)
+-+		freeaddrinfo(res);
+-+
+- 	free(arg->hostname);
+- 	free(arg);
+- 
+--- 
+-2.20.1
+-
+diff --git a/net/fastd/patches/0002-doc-examples-openwrt-fix-init-script-wasn-t-working-.patch b/net/fastd/patches/0002-doc-examples-openwrt-fix-init-script-wasn-t-working-.patch
+deleted file mode 100644
+index b576a987369e93f3cd14fbc83f3c4bffe5cc97d1..0000000000000000000000000000000000000000
+--- a/net/fastd/patches/0002-doc-examples-openwrt-fix-init-script-wasn-t-working-.patch
++++ /dev/null
+@@ -1,29 +0,0 @@
+-From c29b4b0e3cc5bf68129fd0f94f424950b7888deb Mon Sep 17 00:00:00 2001
+-Message-Id: <c29b4b0e3cc5bf68129fd0f94f424950b7888deb.1567630068.git.mschiffer@universe-factory.net>
+-From: Wilfried Klaebe <wklaebe@users.noreply.github.com>
+-Date: Sat, 31 Aug 2019 21:44:13 +0200
+-Subject: [PATCH] doc: examples/openwrt: fix init script, wasn't working with
+- two VPNs
+-
+-If two VPNs were configured via uci, the init script complained about
+-the peer group of its peers not matching its net.
+----
+- doc/examples/openwrt/fastd.init | 2 +-
+- 1 file changed, 1 insertion(+), 1 deletion(-)
+-
+-diff --git a/doc/examples/openwrt/fastd.init b/doc/examples/openwrt/fastd.init
+-index 15737b403ec2..4ba69ece9887 100644
+---- a/doc/examples/openwrt/fastd.init
+-+++ b/doc/examples/openwrt/fastd.init
+-@@ -233,7 +233,7 @@ generate_peer_group_config() {
+- 	config_get group_parent "$group" parent
+- 	[ "$parent" = "$group_parent" ] || return 0
+- 
+--	if [ "$net" != "$peer_net" ]; then
+-+	if [ "$net" != "$group_net" ]; then
+- 		[ -z "$parent" ] || error "warning: the parent of peer group '$group' doesn't match its net, the peer group will be ignored"
+- 		return 0
+- 	fi
+--- 
+-2.23.0
+-
diff --git a/scripts/check_site.lua b/scripts/check_site.lua
index 4e78437a..dd33bf52 100644
--- a/scripts/check_site.lua
+++ b/scripts/check_site.lua
@@ -289,6 +289,19 @@ function M.need_number(path, required)
 	return need_type(path, 'number', required, 'be a number')
 end
 
+function M.need_number_range(path, min, max, required)
+	local val = need_type(path, 'number', required)
+	if not val then
+		return nil
+	end
+
+	if val < min or val > max then
+		var_error(path, val, "be in range [" .. min .. ", " .. max .. "]")
+	end
+
+	return val
+end
+
 function M.need_boolean(path, required)
 	return need_type(path, 'boolean', required, 'be a boolean')
 end
diff --git a/scripts/copy_output.lua b/scripts/copy_output.lua
index e0d4c564..9f14fba5 100755
--- a/scripts/copy_output.lua
+++ b/scripts/copy_output.lua
@@ -1,12 +1,13 @@
 local lib = dofile('scripts/target_lib.lua')
 local env = lib.env
 
+local target = env.GLUON_TARGET
+
+assert(target)
 assert(env.GLUON_IMAGEDIR)
 assert(env.GLUON_PACKAGEDIR)
 
 
-local target = arg[1]
-
 local openwrt_target
 local subtarget = env.SUBTARGET
 if subtarget ~= '' then
@@ -26,33 +27,60 @@ end
 mkdir(env.GLUON_IMAGEDIR..'/factory')
 mkdir(env.GLUON_IMAGEDIR..'/sysupgrade')
 mkdir(env.GLUON_IMAGEDIR..'/other')
+mkdir(env.GLUON_DEBUGDIR)
 
 
 lib.include(target)
 
 
+local function image_source(image)
+	return string.format(
+		'openwrt/bin/targets/%s/openwrt-%s-%s%s%s',
+		bindir, openwrt_target, image.name, image.in_suffix, image.extension)
+end
+
 local function clean(image, name)
 	local dir, file = image:dest_name(name, '\0', '\0')
 	lib.exec {'rm', '-f', dir..'/'..file}
 end
 
-for _, image in ipairs(lib.images) do
-	clean(image, image.image)
+for _, images in pairs(lib.images) do
+	for _, image in ipairs(images) do
+		clean(image, image.image)
 
-	local destdir, destname = image:dest_name(image.image)
-	local source = string.format('openwrt/bin/targets/%s/openwrt-%s-%s%s%s',
-		bindir, openwrt_target, image.name, image.in_suffix, image.extension)
+		local destdir, destname = image:dest_name(image.image)
+		local source = image_source(image)
+
+		lib.exec {'cp', source, destdir..'/'..destname}
 
-	lib.exec {'cp', source, destdir..'/'..destname}
+		for _, alias in ipairs(image.aliases) do
+			clean(image, alias)
 
-	for _, alias in ipairs(image.aliases) do
-		clean(image, alias)
+			local _, aliasname = image:dest_name(alias)
+			lib.exec {'ln', '-s', destname, destdir..'/'..aliasname}
+		end
+	end
 
-		local _, aliasname = image:dest_name(alias)
-		lib.exec {'ln', '-s', destname, destdir..'/'..aliasname}
+	for _, image in ipairs(images) do
+		local source = image_source(image)
+		lib.exec {'rm', '-f', source}
 	end
 end
 
+-- copy kernel image with debug symbols
+local kernel_debug_glob = string.format('%s/gluon-\0-%s-kernel-debug.tar.zst',
+        env.GLUON_DEBUGDIR,
+        target)
+lib.exec {'rm', '-f', kernel_debug_glob}
+local kernel_debug_source = string.format('openwrt/bin/targets/%s/kernel-debug.tar.zst',
+        bindir)
+local kernel_debug_dest = string.format('%s/gluon-%s-%s-%s-kernel-debug.tar.zst',
+        env.GLUON_DEBUGDIR,
+        lib.site_code,
+        env.GLUON_RELEASE,
+        target)
+lib.exec {'cp', kernel_debug_source, kernel_debug_dest}
+
 
 -- Copy opkg repo
 if lib.opkg and (env.GLUON_DEVICES or '') == '' then
diff --git a/scripts/default_feeds.sh b/scripts/default_feeds.sh
index c1871730..0806539a 100644
--- a/scripts/default_feeds.sh
+++ b/scripts/default_feeds.sh
@@ -1 +1,2 @@
-DEFAULT_FEEDS="$(awk '{print $2}' openwrt/feeds.conf.default)"
+# list feeds which don't start with #
+DEFAULT_FEEDS="$(awk '!/^#/ {print $2}' openwrt/feeds.conf.default)"
diff --git a/scripts/feeds.sh b/scripts/feeds.sh
index 7d7d059a..d4b7d6d1 100755
--- a/scripts/feeds.sh
+++ b/scripts/feeds.sh
@@ -11,7 +11,7 @@ rm -rf openwrt/feeds
 rm -rf openwrt/package/feeds
 
 (
-	echo 'src-link gluon_base ../../package'
+	echo "$GLUON_BASE_FEEDS"
 	for feed in $FEEDS; do
 		echo "src-link $feed ../../packages/$feed"
 	done
diff --git a/scripts/generate_manifest.lua b/scripts/generate_manifest.lua
index 73125044..e289412e 100755
--- a/scripts/generate_manifest.lua
+++ b/scripts/generate_manifest.lua
@@ -48,8 +48,10 @@ local function generate(image)
 	end
 end
 
-for _, image in ipairs(lib.images) do
-	if image.subdir == 'sysupgrade' then
-		generate(image)
+for _, images in pairs(lib.images) do
+	for _, image in ipairs(images) do
+		if image.subdir == 'sysupgrade' then
+			generate(image)
+		end
 	end
 end
diff --git a/scripts/target_config.lua b/scripts/target_config.lua
index 452d4f19..28ea4153 100755
--- a/scripts/target_config.lua
+++ b/scripts/target_config.lua
@@ -1,24 +1,5 @@
-local funcs = {}
+local lib = dofile('scripts/target_config_lib.lua')
 
-function funcs.config_message(config, _, ...)
-	config(...)
-end
-
-function funcs.config_package(config, pkg, value)
-	config('CONFIG_PACKAGE_%s=%s', pkg, value)
-end
-
-local lib = dofile('scripts/target_config_lib.lua')(funcs)
-
-
-local output = {}
-
-for config in pairs(lib.configs) do
-	table.insert(output, config)
-end
-
--- The sort will make =y entries override =m ones
-table.sort(output)
-for _, line in ipairs(output) do
-	io.stdout:write(line, '\n')
+for _, config in pairs(lib.configs) do
+	io.stdout:write(config:format(), '\n')
 end
diff --git a/scripts/target_config_check.lua b/scripts/target_config_check.lua
index 160c32a3..373a650e 100755
--- a/scripts/target_config_check.lua
+++ b/scripts/target_config_check.lua
@@ -1,66 +1,47 @@
-local ret = 0
+local errors = false
 
-
-local function fail(...)
-	if ret == 0 then
-		ret = 1
+local function fail(msg)
+	if not errors then
+		errors = true
 		io.stderr:write('Configuration failed:', '\n')
 	end
 
-	io.stderr:write(' * ', string.format(...), '\n')
+	io.stderr:write(' * ', msg, '\n')
 end
 
-local function match_config(f)
-	for line in io.lines('openwrt/.config') do
-		if f(line) then
-			return true
-		end
+local function match_config(expected, actual)
+	if expected == actual then
+		return true
 	end
 
-	return false
-end
-
-local function check_config(pattern)
-	return match_config(function(line) return line == pattern end)
-end
-
-local function check_config_prefix(pattern)
-	return match_config(function(line) return string.sub(line, 1, -2) == pattern end)
-end
-
-
-local funcs = {}
-
-function funcs.config_message(_, message, ...)
-	local pattern = string.format(...)
-
-	if not check_config(pattern) then
-		fail('%s', message)
+	if expected:gsub('=m$', '=y') == actual then
+		return true
 	end
+
+	return false
 end
 
-function funcs.config_package(_, pkg, value)
-	local pattern = string.format('CONFIG_PACKAGE_%s=%s', pkg, value)
-	local res
-	if value == 'y' then
-		res = check_config(pattern)
-	else
-		res = check_config_prefix(string.sub(pattern, 1, -2))
+local function check_config(config)
+	for line in io.lines('openwrt/.config') do
+		if match_config(config, line) then
+			return true
+		end
 	end
 
-	if not res then
-		fail("unable to enable package '%s'", pkg)
-	end
+	return false
 end
 
-local lib = dofile('scripts/target_config_lib.lua')(funcs)
 
-for config, v in pairs(lib.configs) do
-	if v == 2 then
-		if not check_config(config) then
-			fail("unable to set '%s'", config)
+local lib = dofile('scripts/target_config_lib.lua')
+
+for _, config in pairs(lib.configs) do
+	if config.required then
+		if not check_config(config:format()) then
+			fail(config.required)
 		end
 	end
 end
 
-os.exit(ret)
+if errors then
+	os.exit(1)
+end
diff --git a/scripts/target_config_lib.lua b/scripts/target_config_lib.lua
index 71443db7..346a18a7 100644
--- a/scripts/target_config_lib.lua
+++ b/scripts/target_config_lib.lua
@@ -1,82 +1,195 @@
-return function(funcs)
-	local lib = dofile('scripts/target_lib.lua')
-	local env = lib.env
+local lib = dofile('scripts/target_lib.lua')
+local env = lib.env
 
-	assert(env.BOARD)
-	assert(env.SUBTARGET)
+local target = env.GLUON_TARGET
 
-	local target = arg[1]
-	local extra_packages = arg[2]
+assert(target)
+assert(env.BOARD)
+assert(env.SUBTARGET)
 
-	local openwrt_config_target
-	if env.SUBTARGET ~= '' then
-		openwrt_config_target = env.BOARD .. '_' .. env.SUBTARGET
+local openwrt_config_target
+if env.SUBTARGET ~= '' then
+	openwrt_config_target = env.BOARD .. '_' .. env.SUBTARGET
+else
+	openwrt_config_target = env.BOARD
+end
+
+
+-- Split a string into words
+local function split(s)
+	local ret = {}
+	for w in string.gmatch(s, '%S+') do
+		table.insert(ret, w)
+	end
+	return ret
+end
+
+-- Strip leading '-' character
+local function strip_neg(s)
+	if string.sub(s, 1, 1) == '-' then
+		return string.sub(s, 2)
 	else
-		openwrt_config_target = env.BOARD
+		return s
+	end
+end
+
+-- Add an element to a list, removing duplicate entries and handling negative
+-- elements prefixed with a '-'
+local function append_to_list(list, item, keep_neg)
+	local match = strip_neg(item)
+	local ret = {}
+	for _, el in ipairs(list) do
+		if strip_neg(el) ~= match then
+			table.insert(ret, el)
+		end
 	end
+	if keep_neg ~= false or string.sub(item, 1, 1) ~= '-' then
+		table.insert(ret, item)
+	end
+	return ret
+end
+
+local function compact_list(list, keep_neg)
+	local ret = {}
+	for _, el in ipairs(list) do
+		ret  = append_to_list(ret, el, keep_neg)
+	end
+	return ret
+end
 
 
-	local function site_packages(image)
-		return lib.exec_capture_raw(string.format([[
-	MAKEFLAGS= make print _GLUON_IMAGE_=%s --no-print-directory -s -f - <<'END_MAKE'
+local function site_vars(var)
+	return lib.exec_capture_raw(string.format(
+[[
+MAKEFLAGS= make print _GLUON_SITE_VARS_=%s --no-print-directory -s -f - <<'END_MAKE'
 include $(GLUON_SITEDIR)/site.mk
 
 print:
-	echo -n '$(GLUON_$(_GLUON_IMAGE_)_SITE_PACKAGES)'
+	echo -n '$(_GLUON_SITE_VARS_)'
 END_MAKE
-		]], lib.escape(image)))
-	end
+]],
+	lib.escape(var)))
+end
 
-	lib.include('generic')
-	for pkg in string.gmatch(extra_packages, '%S+') do
-		lib.packages {pkg}
+local function site_packages(image)
+	return split(site_vars(string.format('$(GLUON_%s_SITE_PACKAGES)', image)))
+end
+
+-- TODO: Rewrite features.sh in Lua
+local function feature_packages(features)
+	-- Ugly hack: Lua doesn't give us the return code of a popened
+	-- command, so we match on a special __ERROR__ marker
+	local pkgs = lib.exec_capture({'scripts/features.sh', features}, '|| echo __ERROR__')
+	assert(string.find(pkgs, '__ERROR__') == nil, 'Error while evaluating features')
+	return pkgs
+end
+
+-- This involves running lots of processes to evaluate site.mk, so we
+-- add a simple cache
+local class_cache = {}
+local function class_packages(class)
+	if class_cache[class] then
+		return class_cache[class]
 	end
-	lib.include(target)
 
-	lib.check_devices()
+	local features = site_vars(string.format('$(GLUON_FEATURES) $(GLUON_FEATURES_%s)', class))
+	features = table.concat(compact_list(split(features), false), ' ')
 
+	local pkgs = feature_packages(features)
+	pkgs = pkgs .. ' ' .. site_vars(string.format('$(GLUON_SITE_PACKAGES) $(GLUON_SITE_PACKAGES_%s)', class))
 
-	if not lib.opkg then
-		lib.config '# CONFIG_SIGNED_PACKAGES is not set'
-		lib.config 'CONFIG_CLEAN_IPKG=y'
-		lib.packages {'-opkg'}
-	end
+	pkgs = compact_list(split(pkgs))
 
+	class_cache[class] = pkgs
+	return pkgs
+end
+
+local enabled_packages = {}
+-- Arguments: package name and config value (true: y, nil: m, false: unset)
+-- Ensures precedence of y > m > unset
+local function config_package(pkg, v)
+	if v == false then
+		if not enabled_packages[pkg] then
+			lib.try_config('PACKAGE_' .. pkg, false)
+		end
+		return
+	end
 
-	local default_pkgs = ''
-	for _, pkg in ipairs(lib.target_packages) do
-		default_pkgs = default_pkgs .. ' ' .. pkg
+	if v == true or not enabled_packages[pkg] then
+		lib.config('PACKAGE_' .. pkg, v, string.format("unable to enable package '%s'", pkg))
+		enabled_packages[pkg] = true
+	end
+end
 
+local function handle_target_pkgs(pkgs)
+	for _, pkg in ipairs(pkgs) do
 		if string.sub(pkg, 1, 1) == '-' then
-			lib.try_config('# CONFIG_PACKAGE_%s is not set', string.sub(pkg, 2))
+			config_package(string.sub(pkg, 2), false)
 		else
-			funcs.config_package(lib.config, pkg, 'y')
+			config_package(pkg, true)
 		end
 	end
+end
+
+lib.include('generic')
+lib.include(target)
+
+lib.check_devices()
+
+if not lib.opkg then
+	lib.config('SIGNED_PACKAGES', false)
+	lib.config('CLEAN_IPKG', true)
+	lib.config('ALL_NONSHARED', false)
+	lib.packages {'-opkg'}
+end
+
+if #lib.devices > 0 then
+	handle_target_pkgs(lib.target_packages)
 
 	for _, dev in ipairs(lib.devices) do
 		local profile = dev.options.profile or dev.name
-		local device_pkgs = default_pkgs
 
-		local function handle_pkg(pkg)
-			if string.sub(pkg, 1, 1) ~= '-' then
-				funcs.config_package(lib.config, pkg, 'm')
+		local device_pkgs = {}
+		local function handle_pkgs(pkgs)
+			for _, pkg in ipairs(pkgs) do
+				if string.sub(pkg, 1, 1) ~= '-' then
+					config_package(pkg, nil)
+				end
+				device_pkgs = append_to_list(device_pkgs, pkg)
 			end
-			device_pkgs = device_pkgs .. ' ' .. pkg
 		end
 
-		for _, pkg in ipairs(dev.options.packages or {}) do
-			handle_pkg(pkg)
-		end
-		for pkg in string.gmatch(site_packages(dev.image), '%S+') do
-			handle_pkg(pkg)
+		handle_pkgs(lib.target_packages)
+		handle_pkgs(class_packages(dev.options.class))
+		handle_pkgs(dev.options.packages or {})
+		handle_pkgs(site_packages(dev.image))
+
+		local profile_config = string.format('%s_DEVICE_%s', openwrt_config_target, profile)
+		lib.config(
+			'TARGET_DEVICE_' .. profile_config, true,
+			string.format("unable to enable device '%s'", profile)
+		)
+		lib.config(
+			'TARGET_DEVICE_PACKAGES_' .. profile_config,
+			table.concat(device_pkgs, ' ')
+		)
+	end
+else
+	-- x86 fallback: no devices
+	local target_pkgs = {}
+	local function handle_pkgs(pkgs)
+		for _, pkg in ipairs(pkgs) do
+			target_pkgs = append_to_list(target_pkgs, pkg)
 		end
-
-		funcs.config_message(lib.config, string.format("unable to enable device '%s'", profile),
-			'CONFIG_TARGET_DEVICE_%s_DEVICE_%s=y', openwrt_config_target, profile)
-		lib.config('CONFIG_TARGET_DEVICE_PACKAGES_%s_DEVICE_%s="%s"',
-			openwrt_config_target, profile, device_pkgs)
 	end
 
-	return lib
+	-- Just hardcode the class for device-less targets to 'standard'
+	-- - this is x86 only at the moment, and it will have devices
+	-- in OpenWrt 19.07 + 1 as well
+	handle_pkgs(lib.target_packages)
+	handle_pkgs(class_packages('standard'))
+
+	handle_target_pkgs(target_pkgs)
 end
+
+return lib
diff --git a/scripts/target_lib.lua b/scripts/target_lib.lua
index d5133b1e..bb974659 100644
--- a/scripts/target_lib.lua
+++ b/scripts/target_lib.lua
@@ -15,11 +15,6 @@ local env = setmetatable({}, {
 })
 F.env = env
 
-local envtrue = setmetatable({}, {
-	__index = function(_, k) return (tonumber(os.getenv(k)) or 0) > 0 end
-})
-F.envtrue = envtrue
-
 assert(env.GLUON_SITEDIR)
 assert(env.GLUON_TARGETSDIR)
 assert(env.GLUON_RELEASE)
@@ -44,6 +39,7 @@ local default_options = {
 	aliases = {},
 	manifest_aliases = {},
 	packages = {},
+	class = 'standard',
 	deprecated = false,
 	broken = false,
 }
@@ -55,8 +51,12 @@ for dev in string.gmatch(env.GLUON_DEVICES or '', '%S+') do
 	unknown_devices[dev] = true
 end
 
+function F.istrue(v)
+	return (tonumber(v) or 0) > 0
+end
+
 local function want_device(dev, options)
-	if options.broken and not envtrue.BROKEN then
+	if options.broken and not F.istrue(env.BROKEN) then
 		return false
 	end
 	if options.deprecated and env.GLUON_DEPRECATED == '0' then
@@ -98,7 +98,10 @@ function F.escape(s)
 end
 
 local function escape_command(command, raw)
-	local ret = 'exec'
+	local ret = ''
+	if not raw then
+		ret = 'exec'
+	end
 	for _, arg in ipairs(command) do
 		ret = ret .. ' ' .. F.escape(arg)
 	end
@@ -142,17 +145,56 @@ local image_mt = {
 }
 
 local function add_image(image)
-	table.insert(M.images, setmetatable(image, image_mt))
+	local device = image.image
+	M.images[device] = M.images[device] or {}
+	table.insert(M.images[device], setmetatable(image, image_mt))
 end
 
-function F.try_config(...)
-	M.configs[string.format(...)] = 1
+
+local function format_config(k, v)
+	local format
+	if type(v) == 'string' then
+		format = '%s=%q'
+	elseif v == true then
+		format = '%s=y'
+	elseif v == nil then
+		format = '%s=m'
+	elseif v == false then
+		format = '# %s is not set'
+	else
+		format = '%s=%d'
+	end
+	return string.format(format, 'CONFIG_' .. k, v)
 end
 
-function F.config(...)
-	M.configs[string.format(...)] = 2
+local config_mt = {
+	__index = {
+		format = function(config)
+			return format_config(config.key, config.value)
+		end,
+	}
+}
+
+local function do_config(k, v, required)
+	M.configs[k] = setmetatable({
+		key = k,
+		value = v,
+		required = required,
+	}, config_mt)
 end
 
+function F.try_config(k, v)
+	do_config(k, v)
+end
+
+function F.config(k, v, message)
+	if not message then
+		message = string.format("unable to set '%s'", format_config(k, v))
+	end
+	do_config(k, v, message)
+end
+
+
 function F.packages(pkgs)
 	for _, pkg in ipairs(pkgs) do
 		table.insert(M.target_packages, pkg)
@@ -160,6 +202,14 @@ function F.packages(pkgs)
 end
 M.packages = F.packages
 
+local function as_table(v)
+	if type(v) == 'table' then
+		return v
+	else
+		return {v}
+	end
+end
+
 function F.device(image, name, options)
 	options = merge(default_options, options)
 
@@ -191,16 +241,17 @@ function F.device(image, name, options)
 	end
 
 	if options.factory then
-		add_image {
-			image = image,
-			name = name,
-			subdir = 'factory',
-			in_suffix = options.factory,
-			out_suffix = '',
-			extension = options.factory_ext,
-			aliases = options.aliases,
-			manifest_aliases = options.manifest_aliases,
-		}
+		for _, ext in ipairs(as_table(options.factory_ext)) do
+			add_image {
+				image = image,
+				name = name,
+				subdir = 'factory',
+				in_suffix = options.factory,
+				out_suffix = '',
+				extension = ext,
+				aliases = options.aliases,
+			}
+		end
 	end
 	for _, extra_image in ipairs(options.extra_images) do
 		add_image {
@@ -211,7 +262,6 @@ function F.device(image, name, options)
 			out_suffix = extra_image[2],
 			extension = extra_image[3],
 			aliases = options.aliases,
-			manifest_aliases = options.manifest_aliases,
 		}
 	end
 end
@@ -235,7 +285,6 @@ function F.factory_image(image, name, ext, options)
 		out_suffix = '',
 		extension = ext,
 		aliases = options.aliases,
-		manifest_aliases = options.manifest_aliases,
 	}
 end
 
diff --git a/scripts/update-patches.sh b/scripts/update-patches.sh
index 94f103fa..939c1640 100755
--- a/scripts/update-patches.sh
+++ b/scripts/update-patches.sh
@@ -21,6 +21,6 @@ for module in $GLUON_MODULES; do
 	for commit in $(git rev-list --reverse --no-merges base..patched); do
 		(( ++n ))
 		mkdir -p "${GLUON_PATCHESDIR}/$module"
-		git -c core.abbrev=40 show --pretty=format:'From: %an <%ae>%nDate: %aD%nSubject: %B' --no-renames --binary "$commit" > "${GLUON_PATCHESDIR}/$module/$(printf '%04u' $n)-$(git show -s --pretty=format:%f "$commit").patch"
+		git -c core.abbrev=40 show --pretty=format:'From: %an <%ae>%nDate: %aD%nSubject: %B' --no-renames --binary "$commit" > "${GLUON_PATCHESDIR}/$module/$(printf '%04u' "$n")-$(git show -s --pretty=format:%f "$commit").patch"
 	done
 done
diff --git a/targets/ar71xx-generic b/targets/ar71xx-generic
index d0920d73..363ad019 100644
--- a/targets/ar71xx-generic
+++ b/targets/ar71xx-generic
@@ -1,5 +1,5 @@
-config 'CONFIG_GLUON_SPECIALIZE_KERNEL=y'
-config 'CONFIG_TARGET_SQUASHFS_BLOCK_SIZE=64'
+config('GLUON_SPECIALIZE_KERNEL', true)
+config('TARGET_SQUASHFS_BLOCK_SIZE', 64)
 
 local ATH10K_PACKAGES = {
 	'kmod-ath10k',
@@ -35,6 +35,7 @@ device('8devices-carambola2-board', 'carambola2', {
 
 device('alfa-network-ap121f', 'ap121f', {
 	factory = false,
+	class = 'tiny', -- 32M ath9k
 })
 
 
@@ -99,6 +100,7 @@ device('d-link-dir-505-rev-a1', 'dir-505-a1', {
 device('d-link-dir-825-rev-b1', 'dir-825-b1', {
 	profile = 'DIR825B1',
 	factory = false,
+	class = 'tiny', -- Only 6M of usable Firmware space
 })
 
 
@@ -129,6 +131,7 @@ device('gl.inet-gl-ar750', 'gl-ar750', {
 
 device('linksys-wrt160nl', 'wrt160nl', {
 	profile = 'WRT160NL',
+	class = 'tiny', -- 32M ath9k
 })
 
 
@@ -294,23 +297,28 @@ device('tp-link-wbs510-v1', 'wbs510-v1', {
 })
 
 device('tp-link-tl-wr710n-v1', 'tl-wr710n-v1', {
+	class = 'tiny', -- 32M ath9k
 	packages = { 'zram-swap' },
 })
 device('tp-link-tl-wr710n-v2.1', 'tl-wr710n-v2.1', {
+	class = 'tiny', -- 32M ath9k
 	packages = { 'zram-swap' },
 })
 
 device('tp-link-tl-wr810n-v1', 'tl-wr810n-v1')
 
 device('tp-link-tl-wr842n-nd-v1', 'tl-wr842n-v1', {
+	class = 'tiny', -- 32M ath9k
 	packages = { 'zram-swap' },
 })
 device('tp-link-tl-wr842n-nd-v2', 'tl-wr842n-v2', {
+	class = 'tiny', -- 32M ath9k
 	packages = { 'zram-swap' },
 })
 device('tp-link-tl-wr842n-nd-v3', 'tl-wr842n-v3')
 
 device('tp-link-tl-wr1043n-nd-v1', 'tl-wr1043nd-v1', {
+	class = 'tiny', -- 32M ath9k
 	packages = { 'zram-swap' },
 })
 device('tp-link-tl-wr1043n-nd-v2', 'tl-wr1043nd-v2')
@@ -344,11 +352,13 @@ device('tp-link-archer-c7-v5', 'archer-c7-v5', {
 device('tp-link-archer-c25-v1', 'archer-c25-v1', {
 	packages = ATH10K_PACKAGES_QCA9887,
 	broken = true, -- OOM with 5GHz enabled in most environments
+	class = 'tiny', -- 64M ath9k + ath10k
 })
 
 device('tp-link-archer-c58-v1', 'archer-c58-v1', {
 	packages = ATH10K_PACKAGES_QCA9888,
 	broken = true, -- OOM with 5GHz enabled in most environments
+	class = 'tiny', -- 64M ath9k + ath10k
 })
 
 device('tp-link-archer-c59-v1', 'archer-c59-v1', {
@@ -358,25 +368,30 @@ device('tp-link-archer-c59-v1', 'archer-c59-v1', {
 device('tp-link-archer-c60-v1', 'archer-c60-v1', {
 	packages = ATH10K_PACKAGES_QCA9888,
 	broken = true, -- OOM with 5GHz enabled in most environments
+	class = 'tiny', -- 64M ath9k + ath10k
 })
 
 device('tp-link-archer-c60-v2', 'archer-c60-v2', {
 	packages = ATH10K_PACKAGES_QCA9888,
 	broken = true, -- OOM with 5GHz enabled in most environments
+	class = 'tiny', -- 64M ath9k + ath10k
 })
 
 device('tp-link-re355', 're355-v1', {
 	packages = ATH10K_PACKAGES,
 	broken = true, -- OOM with 5GHz enabled in most environments if device is 64M RAM variant
+	class = 'tiny', -- Only 6M of usable Firmware space
 })
 
 device('tp-link-tl-wr902ac-v1', 'tl-wr902ac-v1', {
 	packages = ATH10K_PACKAGES_QCA9887,
 	broken = true, -- OOM due to insufficient RAM for ath10k expected
+	class = 'tiny', -- 64M ath9k + ath10k
 })
 
 device('tp-link-re450', 're450-v1', {
 	packages = ATH10K_PACKAGES,
+	class = 'tiny', -- Only 6M of usable Firmware space
 })
 
 
@@ -384,11 +399,16 @@ device('tp-link-re450', 're450-v1', {
 
 device('ubiquiti-airgateway', 'ubnt-air-gateway', {
 	aliases = {'ubiquiti-airgateway-lr'},
+	class = 'tiny', -- 32M ath9k
 })
 
-device('ubiquiti-airgateway-pro', 'ubnt-air-gateway-pro')
+device('ubiquiti-airgateway-pro', 'ubnt-air-gateway-pro', {
+	class = 'tiny', -- 32M ath9k
+})
 
-device('ubiquiti-airrouter', 'ubnt-airrouter')
+device('ubiquiti-airrouter', 'ubnt-airrouter', {
+	class = 'tiny', -- 32M ath9k
+})
 
 device('ubiquiti-bullet-m', 'ubnt-bullet-m', {
 	aliases = {
@@ -399,6 +419,7 @@ device('ubiquiti-bullet-m', 'ubnt-bullet-m', {
 		'ubiquiti-picostation-m2',
 	},
 	packages = { 'zram-swap' },
+	class = 'tiny', -- 32M ath9k
 })
 
 device('ubiquiti-rocket-m', 'ubnt-rocket-m', {
@@ -414,6 +435,7 @@ device('ubiquiti-nanostation-m', 'ubnt-nano-m', {
 		'ubiquiti-nanostation-m5',
 	},
 	packages = { 'zram-swap' },
+	class = 'tiny', -- 32M ath9k
 })
 
 device('ubiquiti-loco-m-xw', 'ubnt-loco-m-xw', {
@@ -458,6 +480,7 @@ device('ubiquiti-unifiap-outdoor+', 'ubnt-unifi-outdoor-plus')
 
 device('ubiquiti-ls-sr71', 'ubnt-ls-sr71', {
 	broken = true, -- untested
+	class = 'tiny', -- 32M ath9k
 })
 
 device('ubiquiti-unifi-ac-lite', 'ubnt-unifiac-lite', {
diff --git a/targets/ar71xx-mikrotik b/targets/ar71xx-mikrotik
index 6a783b5e..81a053f2 100644
--- a/targets/ar71xx-mikrotik
+++ b/targets/ar71xx-mikrotik
@@ -1,4 +1,4 @@
-config 'CONFIG_GLUON_SPECIALIZE_KERNEL=y'
+config('GLUON_SPECIALIZE_KERNEL', true)
 
 defaults {
 	factory = false,
diff --git a/targets/ar71xx-nand b/targets/ar71xx-nand
index e52b82e0..89afd52c 100644
--- a/targets/ar71xx-nand
+++ b/targets/ar71xx-nand
@@ -1,4 +1,4 @@
-config 'CONFIG_GLUON_SPECIALIZE_KERNEL=y'
+config('GLUON_SPECIALIZE_KERNEL', true)
 
 local ATH10K_PACKAGES = {'kmod-ath10k', '-kmod-ath10k-ct', 'ath10k-firmware-qca988x', '-ath10k-firmware-qca988x-ct'}
 
diff --git a/targets/ar71xx-tiny b/targets/ar71xx-tiny
index 1d09fc17..a0a8d510 100644
--- a/targets/ar71xx-tiny
+++ b/targets/ar71xx-tiny
@@ -1,4 +1,4 @@
-config 'CONFIG_GLUON_SPECIALIZE_KERNEL=y'
+config('GLUON_SPECIALIZE_KERNEL', true)
 
 no_opkg()
 
@@ -12,6 +12,7 @@ packages {
 
 defaults {
 	deprecated = true, -- 4/32
+	class = 'tiny',
 }
 
 
diff --git a/targets/ath79-generic b/targets/ath79-generic
index 7caf6578..c30d24a2 100644
--- a/targets/ath79-generic
+++ b/targets/ath79-generic
@@ -6,6 +6,14 @@ local ATH10K_PACKAGES_QCA9880 = {
 	'-ath10k-firmware-qca988x-ct',
 }
 
+local ATH10K_PACKAGES_QCA9887 = {
+	'kmod-ath10k',
+	'-kmod-ath10k-ct',
+	'-kmod-ath10k-ct-smallbuffers',
+	'ath10k-firmware-qca9887',
+	'-ath10k-firmware-qca9887-ct',
+}
+
 local ATH10K_PACKAGES_QCA9888 = {
 	'kmod-ath10k',
 	'-kmod-ath10k-ct',
@@ -49,7 +57,12 @@ device('devolo-wifi-pro-1750x', 'devolo_dvl1750x', {
 -- GL.iNet
 
 device('gl.inet-gl-ar300m-lite', 'glinet_gl-ar300m-lite', {
-    factory = false,
+	factory = false,
+})
+
+device('gl.inet-gl-ar750s-nor', 'glinet_gl-ar750s', {
+	factory = false,
+	packages = ATH10K_PACKAGES_QCA9887,
 })
 
 -- OCEDO
@@ -64,3 +77,4 @@ device('tp-link-archer-c6-v2', 'tplink_archer-c6-v2', {
 	packages = ATH10K_PACKAGES_QCA9888,
 })
 
+device('tp-link-cpe220-v3', 'tplink_cpe220-v3')
diff --git a/targets/generic.2020.2 b/targets/generic.2020.2
new file mode 100644
index 00000000..89aba6f2
--- /dev/null
+++ b/targets/generic.2020.2
@@ -0,0 +1,79 @@
+assert(env.GLUON_LANGS)
+
+
+config('GLUON_SITEDIR', env.GLUON_SITEDIR)
+config('GLUON_RELEASE', env.GLUON_RELEASE)
+try_config('GLUON_BRANCH', env.GLUON_BRANCH or '')
+
+for lang in string.gmatch(env.GLUON_LANGS, '%S+') do
+	try_config('GLUON_WEB_LANG_' .. lang, true)
+end
+
+config('TARGET_' .. env.BOARD, true)
+if env.SUBTARGET ~= '' then
+	config(string.format('TARGET_%s_%s', env.BOARD, env.SUBTARGET), true)
+end
+
+-- Disable non-default feeds in distfeeds.conf
+config('FEED_gluon_base', false)
+
+local default_feeds = {}
+for feed in string.gmatch(exec_capture_raw('. scripts/default_feeds.sh && echo "$DEFAULT_FEEDS"'), '%S+') do
+	default_feeds[feed] = true
+end
+
+for feed in string.gmatch(exec_capture_raw('. scripts/modules.sh && echo -n "$FEEDS"'), '%S+') do
+	if not default_feeds[feed] then
+		config('FEED_' .. feed, false)
+	end
+end
+
+
+config('TARGET_ROOTFS_INITRAMFS', false)
+
+config('DEVEL', true)
+config('ALL_NONSHARED', true)
+
+config('PACKAGE_usbip', false) -- fails to build
+config('PACKAGE_kmod-jool', false) -- fails to build
+
+config('BUSYBOX_CUSTOM', true)
+config('BUSYBOX_CONFIG_FEATURE_PREFER_IPV4_ADDRESS', false)
+
+try_config('PACKAGE_ATH_DEBUG', true)
+
+try_config('TARGET_SQUASHFS_BLOCK_SIZE', 256)
+
+config('KERNEL_IP_MROUTE', false)
+config('KERNEL_IPV6_MROUTE', false)
+
+config('COLLECT_KERNEL_DEBUG', true)
+
+try_config('TARGET_MULTI_PROFILE', true)
+try_config('TARGET_PER_DEVICE_ROOTFS', true)
+
+config('GLUON_MULTIDOMAIN', istrue(env.GLUON_MULTIDOMAIN))
+
+config('AUTOREMOVE', istrue(env.GLUON_AUTOREMOVE))
+
+if istrue(env.GLUON_DEBUG) then
+	config('DEBUG', true)
+	config('NO_STRIP', true)
+	config('USE_STRIP', false)
+	config('USE_SSTRIP', false)
+
+	try_config('TARGET_ROOTFS_PARTSIZE', 500)
+end
+
+config('GLUON_MINIFY', istrue(env.GLUON_MINIFY))
+
+packages {
+	'-kmod-ipt-offload',
+	'-odhcpd-ipv6only',
+	'-ppp',
+	'-ppp-mod-pppoe',
+	'-wpad-mini',
+	'-wpad-basic',
+	'gluon-core',
+	'ip6tables',
+}
diff --git a/targets/ipq40xx-generic b/targets/ipq40xx-generic
index 8df11ea2..0dca0426 100644
--- a/targets/ipq40xx-generic
+++ b/targets/ipq40xx-generic
@@ -43,6 +43,17 @@ device('avm-fritz-repeater-1200', 'avm_fritzrepeater-1200', {
 })
 
 
+-- EnGenius
+
+device('engenius-ens620ext', 'engenius_ens620ext', {
+	factory = false,
+	extra_images = {
+		{'-squashfs-factory_30', '-factory_fw30', '.bin'},
+		{'-squashfs-factory_35', '-factory_fw35', '.bin'},
+	},
+})
+
+
 -- GL.iNet
 
 device('gl.inet-gl-b1300', 'glinet_gl-b1300', {
@@ -50,6 +61,11 @@ device('gl.inet-gl-b1300', 'glinet_gl-b1300', {
 })
 
 
+-- Linksys
+
+device('linksys-ea6350v3', 'linksys_ea6350v3')
+
+
 -- NETGEAR
 
 device('netgear-ex6100v2', 'netgear_ex6100v2', {
@@ -76,6 +92,7 @@ device('zyxel-nbg6617', 'zyxel_nbg6617')
 
 device('zyxel-wre6606', 'zyxel_wre6606', {
 	factory = false,
+	class = 'tiny', -- 128M ath10k + ath10k
 })
 
 
diff --git a/targets/lantiq-xrx200 b/targets/lantiq-xrx200
index 43526eb0..f015fb28 100644
--- a/targets/lantiq-xrx200
+++ b/targets/lantiq-xrx200
@@ -1,12 +1,43 @@
+packages {
+	'-ltq-vdsl-vr9-vectoring-fw-installer',
+	'-kmod-ltq-vdsl-vr9-mei',
+	'-kmod-ltq-vdsl-vr9',
+	'-kmod-ltq-atm-vr9',
+	'-kmod-ltq-ptm-vr9',
+	'-kmod-ltq-deu-vr9',
+	'-ltq-vdsl-app',
+	'-dsl-vrx200-firmware-xdsl-a',
+	'-dsl-vrx200-firmware-xdsl-b-patch',
+	'-ppp-mod-pppoa',
+}
+
+
 device('avm-fritz-box-7360-sl', 'avm_fritz7360sl', {
-        factory = false,
-        aliases = {'avm-fritz-box-7360-v1', 'avm-fritz-box-7360-v2'},
+	factory = false,
+	aliases = {'avm-fritz-box-7360-v1', 'avm-fritz-box-7360-v2'},
 })
 
 device('avm-fritz-box-7362-sl', 'avm_fritz7362sl', {
-        factory = false,
+	factory = false,
 })
 
 device('avm-fritz-box-7412', 'avm_fritz7412', {
-        factory = false,
+	factory = false,
+})
+
+-- TP-Link
+
+  -- CAVEAT: These devices don't have a dedicated WAN port.
+  --         All ethernet ports on the device are bridged and
+  --         used as WAN ports. Clients connected to these
+  --         ports will be connected to your private network.
+
+device('tp-link-td-w8970', 'tplink_tdw8970', {
+	factory = false,
+})
+
+device('tp-link-td-w8980', 'tplink_tdw8980', {
+	factory = false,
+	aliases = {'tp-link-td-w9980'},
+	broken = true, -- 5GHz unsupported
 })
diff --git a/targets/lantiq-xway b/targets/lantiq-xway
index c75f630a..bfbc195a 100644
--- a/targets/lantiq-xway
+++ b/targets/lantiq-xway
@@ -2,3 +2,6 @@ device('avm-fritz-box-7312', 'avm_fritz7312', {
 	factory = false,
 })
 
+device('netgear-dgn3500b', 'netgear_dgn3500b', {
+	factory_ext = '.img',
+})
diff --git a/targets/mpc85xx-p1020 b/targets/mpc85xx-p1020
index ee120111..952b9998 100644
--- a/targets/mpc85xx-p1020
+++ b/targets/mpc85xx-p1020
@@ -1,7 +1,7 @@
 -- Aerohive
 
 device('aerohive-hiveap-330', 'hiveap-330', {
-        factory = false,
+	factory = false,
 })
 
 
@@ -15,6 +15,6 @@ device('enterasys-ws-ap3710i', 'ws-ap3710i', {
 -- OCEDO
 
 device('ocedo-panda', 'panda', {
-        factory = false,
+	factory = false,
 })
 
diff --git a/targets/ramips-mt7620 b/targets/ramips-mt7620
index 9dd6a2d9..ff2e4529 100644
--- a/targets/ramips-mt7620
+++ b/targets/ramips-mt7620
@@ -35,11 +35,11 @@ device('nexx-wt3020-8m', 'wt3020-8M', {
 
 local tplink_region_suffix = ''
 if (env.GLUON_REGION or '') ~= '' then
-        tplink_region_suffix = '-' .. env.GLUON_REGION
+	tplink_region_suffix = '-' .. env.GLUON_REGION
 end
 
 device('tp-link-archer-c2-v1', 'tplink_c2-v1', {
-        factory = false,
+	factory = false,
 })
 
 device('tp-link-archer-c20-v1', 'tplink_c20-v1')
@@ -47,7 +47,7 @@ device('tp-link-archer-c20-v1', 'tplink_c20-v1')
 device('tp-link-archer-c20i', 'ArcherC20i')
 
 device('tp-link-archer-c50', 'ArcherC50v1', {
-        factory = '-squashfs-factory' .. tplink_region_suffix,
+	factory = '-squashfs-factory' .. tplink_region_suffix,
 })
 
 
diff --git a/targets/ramips-mt7621 b/targets/ramips-mt7621
index 9b3ddf86..978eb619 100644
--- a/targets/ramips-mt7621
+++ b/targets/ramips-mt7621
@@ -13,7 +13,7 @@ device('d-link-dir-860l-b1', 'dir-860l-b1')
 -- Netgear
 
 device('netgear-ex6150', 'netgear_ex6150', {
-        factory_ext = '.chk',
+	factory_ext = '.chk',
 })
 
 device('netgear-r6220', 'r6220', {
diff --git a/targets/ramips-mt76x8 b/targets/ramips-mt76x8
index c1e0354f..ccc83143 100644
--- a/targets/ramips-mt76x8
+++ b/targets/ramips-mt76x8
@@ -1,3 +1,8 @@
+-- Cudy
+
+device('cudy-wr1000', 'cudy_wr1000')
+
+
 -- GL.iNet
 
 device('gl-mt300n-v2', 'gl-mt300n-v2', {
diff --git a/targets/ramips-rt305x b/targets/ramips-rt305x
index d8131f75..8678f12e 100644
--- a/targets/ramips-rt305x
+++ b/targets/ramips-rt305x
@@ -1,8 +1,12 @@
-config '# CONFIG_KERNEL_KALLSYMS is not set'
-config 'CONFIG_GLUON_SPECIALIZE_KERNEL=y'
+config('KERNEL_KALLSYMS', false)
+config('GLUON_SPECIALIZE_KERNEL', true)
 
 no_opkg()
 
+defaults {
+	class = 'tiny',  -- 32M RAM
+}
+
 local STRIP_USB_PACKAGES = {
 	'-kmod-usb-core',
 	'-kmod-usb-dwc2',
diff --git a/targets/targets.mk b/targets/targets.mk
index 9009898e..bad38ac2 100644
--- a/targets/targets.mk
+++ b/targets/targets.mk
@@ -19,6 +19,7 @@ $(eval $(call GluonTarget,ramips,rt305x))
 $(eval $(call GluonTarget,sunxi,cortexa7))
 $(eval $(call GluonTarget,x86,generic))
 $(eval $(call GluonTarget,x86,geode))
+$(eval $(call GluonTarget,x86,legacy))
 $(eval $(call GluonTarget,x86,64))
 
 
diff --git a/targets/x86-legacy b/targets/x86-legacy
new file mode 100644
index 00000000..79667dad
--- /dev/null
+++ b/targets/x86-legacy
@@ -0,0 +1,10 @@
+include 'x86.inc'
+
+packages {
+        '-kmod-gpio-nct5104d',
+        '-kmod-leds-gpio',
+        '-kmod-leds-apu2',
+}
+
+factory_image('x86-legacy', 'combined-squashfs', '.img.gz')
+sysupgrade_image('x86-legacy', 'combined-squashfs', '.img.gz')
diff --git a/targets/x86.inc b/targets/x86.inc
index 4ca2bb8c..c6037c68 100644
--- a/targets/x86.inc
+++ b/targets/x86.inc
@@ -1,5 +1,5 @@
-config 'CONFIG_VDI_IMAGES=y'
-config 'CONFIG_VMDK_IMAGES=y'
+config('VDI_IMAGES', true)
+config('VMDK_IMAGES', true)
 
 packages {
 	'kmod-3c59x',
@@ -34,5 +34,11 @@ packages {
 	'kmod-usb-serial',
 	'kmod-ath10k',
 	'ath10k-firmware-qca9887',
+	'ath10k-firmware-qca9888',
 	'ath10k-firmware-qca988x',
+	'ath10k-firmware-qca9984',
+	'kmod-mt76x0e',
+	'kmod-mt76x2',
+	'kmod-mt7603',
+	'kmod-mt7615e',
 }
