--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -53,16 +53,15 @@ local function is_outdoor()
 end

 local function get_channel(radio, config)
-       local channel
-       if wireless.preserve_channels(uci) then
-               -- preserved channel always wins
-               channel = radio.channel
-       elseif  (radio.hwmode == '11a' or radio.hwmode == '11na') and is_outdoor() then
-               -- actual channel will be picked and probed from chanlist
-               channel = 'auto'
-       end
-
-       return channel or config.channel()
+        local channel
+       if (radio.hwmode == '11a' or radio.hwmode == '11na') and is_outdoor() and not uci:get_first('gluon-core', 'wireless', 'force_channel_outdoor') then
+                -- actual channel will be picked and probed from chanlist
+                channel = 'auto'
+       elseif  uci:get_first('gluon-core', 'wireless', 'preserve_channels') then
+               -- preserved channel always wins only indoors
+               channel = radio.channel
+        end
+        return channel or config.channel()
 end

 local function get_htmode(radio)

