diff --git a/package/gluon-core/files/lib/gluon/upgrade/200-wireless b/package/gluon-core/files/lib/gluon/upgrade/200-wireless
index b9495d8..462d552 100755
--- a/package/gluon-core/files/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/files/lib/gluon/upgrade/200-wireless
@@ -8,13 +8,12 @@ local sysconfig = require 'gluon.sysconfig'
 -- Initial
 if not sysconfig.gluon_version then
   uci:delete_all('wireless', 'wifi-iface')
+  uci:set('wireless', radio, 'channel', config.channel)
 end

 local function configure_radio(radio, index, config)
   if config then
     uci:delete('wireless', radio, 'disabled')
-
-    uci:set('wireless', radio, 'channel', config.channel)
     uci:set('wireless', radio, 'htmode', 'HT20')
     uci:set('wireless', radio, 'country', site.regdom)
   end
